Add basic support to play or pause media playback for active (Now Playing) media

whsieh opened the pull request:
#### f89ad9045bd3cacee1f67b30d07a8420fb563be1
<pre>
Add basic support to play or pause media playback for active (Now Playing) media
<a href="https://bugs.webkit.org/show_bug.cgi?id=276235">https://bugs.webkit.org/show_bug.cgi?id=276235</a>
<a href="https://rdar.apple.com/131137466">rdar://131137466</a>

Reviewed by Eric Carlson.

Add support for `-_playPredominantOrNowPlayingMediaSession:` and `-_pauseNowPlayingMediaSession:`.
See comments below.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _playPredominantOrNowPlayingMediaSession:]):
(-[WKWebView _pauseNowPlayingMediaSession:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::playPredominantOrNowPlayingMediaSession):

Add plumbing to get to the playback session model of the active playback session. Note that these
currently take completion handlers (despite being synchronous), since the next patch will add a
heuristic to asynchronously find and play the most prominent video element in the viewport as a
fallback, in the case where there&apos;s no current media session.

(WebKit::WebPageProxy::pauseNowPlayingMediaSession):
(WebKit::WebPageProxy::tryToSendCommandToActiveControlledVideo):

Add a helper method to send a `RemoteControlCommandType` to the current playback session model.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/VideoControlsManager.mm:
(-[VideoControlsManagerTestWebView _playPredominantOrNowPlayingMediaSession]):
(-[VideoControlsManagerTestWebView _pauseNowPlayingMediaSession]):
(-[VideoControlsManagerTestWebView waitForVideoToPlay]):
(-[VideoControlsManagerTestWebView waitForVideoToPause]):
(TestWebKitAPI::TEST(VideoControlsManager, TogglePlaybackForControlledVideo)):

Add an API test that exercises the new `WKWebView` methods.

Canonical link: <a href="https://commits.webkit.org/280686@main">https://commits.webkit.org/280686@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dabcc16e2560727cdebdcfd71ac7568842c55032

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57286 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36614 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9761 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60908 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7729 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59414 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44238 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7919 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46382 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5451 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59316 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34358 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49465 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27245 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31140 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6778 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6734 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7049 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62587 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1199 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7143 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53643 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1204 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49501 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53719 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12669 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1015 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32443 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33528 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34613 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33274 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d5e9c33a4c96d79d4431d36dfae4f0d5cd9c6088. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/30500

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/dabcc16e2560727cdebdcfd71ac7568842c55032. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/30500

whsieh commented:
Thanks for the review!

whsieh committed:
Add basic support to play or pause media playback for active (Now Playing) media
https://bugs.webkit.org/show_bug.cgi?id=276235
rdar://131137466

Reviewed by Eric Carlson.

Add support for `-_playPredominantOrNowPlayingMediaSession:` and `-_pauseNowPlayingMediaSession:`.
See comments below.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _playPredominantOrNowPlayingMediaSession:]):
(-[WKWebView _pauseNowPlayingMediaSession:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::playPredominantOrNowPlayingMediaSession):

Add plumbing to get to the playback session model of the active playback session. Note that these
currently take completion handlers (despite being synchronous), since the next patch will add a
heuristic to asynchronously find and play the most prominent video element in the viewport as a
fallback, in the case where there's no current media session.

(WebKit::WebPageProxy::pauseNowPlayingMediaSession):
(WebKit::WebPageProxy::tryToSendCommandToActiveControlledVideo):

Add a helper method to send a `RemoteControlCommandType` to the current playback session model.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/VideoControlsManager.mm:
(-[VideoControlsManagerTestWebView _playPredominantOrNowPlayingMediaSession]):
(-[VideoControlsManagerTestWebView _pauseNowPlayingMediaSession]):
(-[VideoControlsManagerTestWebView waitForVideoToPlay]):
(-[VideoControlsManagerTestWebView waitForVideoToPause]):
(TestWebKitAPI::TEST(VideoControlsManager, TogglePlaybackForControlledVideo)):

Add an API test that exercises the new `WKWebView` methods.

Canonical link: https://commits.webkit.org/280686@main

webkit-commit-queue commented:
Committed 280686@main (f89ad9045bd3): <https://commits.webkit.org/280686@main>

Reviewed commits have been landed. Closing PR #30500 and removing active labels.
